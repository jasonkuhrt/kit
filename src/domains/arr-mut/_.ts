export * as ArrMut from './__.js'
