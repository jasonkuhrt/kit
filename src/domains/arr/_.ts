// @ts-expect-error Duplicate identifier
export * as Arr from './__.js'

/**
 * Array utilities for working with readonly and mutable arrays.
 *
 * Provides functional utilities for array operations including mapping, filtering,
 * type guards, and conversions. Emphasizes immutable operations and type safety.
 *
 * @category Domains
 */
export namespace Arr {}
