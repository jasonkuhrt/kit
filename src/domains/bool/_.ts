export * as Bool from './__.js'
