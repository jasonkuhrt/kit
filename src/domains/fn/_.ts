// @ts-expect-error Duplicate identifier
export * as Fn from './__.js'

/**
 * Function utilities for functional programming patterns.
 *
 * Provides utilities for function composition, currying, partial application,
 * piping, and common functional patterns like identity and constant functions.
 * Includes function analysis tools and endomorphism utilities.
 *
 * @category Domains
 */
export namespace Fn {}

// Re-export Kind at top level for easier access
export * as Kind from './kind.js'
