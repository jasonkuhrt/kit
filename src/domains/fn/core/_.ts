export * as CoreFn from './__.js'
