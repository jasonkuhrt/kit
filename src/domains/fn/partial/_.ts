export * as Partial from './__.js'
