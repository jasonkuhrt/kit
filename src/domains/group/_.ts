export * as Group from './__.js'
