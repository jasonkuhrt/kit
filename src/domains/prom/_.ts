// @ts-expect-error Duplicate identifier
export * as Prom from './__.js'

/**
 * Promise utilities for asynchronous operations.
 *
 * Provides utilities for working with Promises including deferred promise
 * creation, promise combinators, and async control flow patterns.
 *
 * @category Domains
 */
export namespace Prom {}
