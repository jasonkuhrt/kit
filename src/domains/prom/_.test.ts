import { Test } from '#test'
import { property } from '#test/test'
import fc from 'fast-check'
import { describe, expect, test } from 'vitest'
import { Prom } from './_.js'

describe('isShape', () => {
  // test('detects promises and thenables', () => {
  //   expect(Prom.isShape(Promise.resolve(42))).toBe(true)
  //   // eslint-disable-next-line unicorn/no-thenable
  //   expect(Prom.isShape({ then: () => {}, catch: () => {}, finally: () => {} })).toBe(true)
  //   // eslint-disable-next-line unicorn/no-thenable
  //   expect(Prom.isShape({ then: () => {} })).toBe(false)
  //   // eslint-disable-next-line unicorn/no-thenable
  //   expect(Prom.isShape({ then: 'not a function', catch: () => {}, finally: () => {} })).toBe(false)
  // })
  Test
    .on(Prom.isShape)
    .describe('true', [
      [[Promise.resolve(42)], true],
      [[{ then() {}, catch() {}, finally() {} }], true],
    ])
    .describe('false', [
      [[{ then() {} }], false],
      [[{ then: 'not a function', catch() {}, finally() {} }], false],
    ])
    .test()

  property(
    'rejects non-promise values',
    fc.oneof(fc.integer(), fc.string(), fc.boolean(), fc.constant(null), fc.constant(undefined), fc.object()),
    (value) => {
      expect(Prom.isShape(value)).toBe(false)
    },
  )
})

describe('createDeferred', () => {
  Test
    .describe('state tracking')
    .onSetup(() => ({
      d: Prom.createDeferred<number>(),
    }))
    .outputDefault(({ d }) => d)
    .case('initial', () => {})
    .case('after resolve', ({ d }) => {
      d.resolve(1)
    })
    .case('after reject', ({ d }) => {
      d.reject(new Error('test'))
      d.promise.catch(() => {}) // Prevent unhandled rejection
    })
    .test()

  Test
    .describe('strict')
    .onSetup(() => ({
      d: Prom.createDeferred<number>({ strict: true }),
    }))
    .outputDefault(({ d }) => d)
    .case('resolve resolve calls throw', ({ d }) => {
      d.resolve(1)
      d.resolve(1)
    })
    .case('resolve reject calls throw', ({ d }) => {
      d.reject(new Error('test'))
      d.promise.catch(() => {}) // Prevent unhandled rejection
      d.resolve(1)
    })
    .case('reject resolve calls throw', ({ d }) => {
      d.reject(new Error('test'))
      d.promise.catch(() => {}) // Prevent unhandled rejection
      d.resolve(1)
    })
    .test()

  test('destructuring', () => {
    const d = Prom.createDeferred<number>()
    const { isResolved, isRejected, isSettled } = d
    expect({ isResolved, isRejected, isSettled }).toEqual({ isResolved: false, isRejected: false, isSettled: false })
  })
})
