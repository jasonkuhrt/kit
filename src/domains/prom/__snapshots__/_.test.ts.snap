// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`createDeferred > state tracking > after reject 1`] = `
"
╔══════════════════════════════════════════════════╗ RUNNER
d.reject(new Error("test"));
d.promise.catch(() => {
});
╠══════════════════════════════════════════════════╣ OUTPUTS RETURN OBJECT
{
  promise: Promise [Promise] {},
  resolve: [Function: resolve],
  reject: [Function: reject],
  isResolved: false,
  isRejected: true,
  isSettled: true
}
╚══════════════════════════════════════════════════╝"
`;

exports[`createDeferred > state tracking > after resolve 1`] = `
"
╔══════════════════════════════════════════════════╗ RUNNER
d.resolve(1);
╠══════════════════════════════════════════════════╣ OUTPUTS RETURN OBJECT
{
  promise: Promise [Promise] {},
  resolve: [Function: resolve],
  reject: [Function: reject],
  isResolved: true,
  isRejected: false,
  isSettled: true
}
╚══════════════════════════════════════════════════╝"
`;

exports[`createDeferred > state tracking > initial 1`] = `
"
╔══════════════════════════════════════════════════╗ RUNNER

╠══════════════════════════════════════════════════╣ OUTPUTS RETURN OBJECT
{
  promise: Promise [Promise] {},
  resolve: [Function: resolve],
  reject: [Function: reject],
  isResolved: false,
  isRejected: false,
  isSettled: false
}
╚══════════════════════════════════════════════════╝"
`;

exports[`createDeferred > strict > reject resolve calls throw 1`] = `
"
╔══════════════════════════════════════════════════╗ RUNNER
d.reject(new Error("test"));
d.promise.catch(() => {
});
d.resolve(1);
╠══════════════════════════════════════════════════╣ OUTPUTS THROW ERROR
Error: Deferred promise already settled
╚══════════════════════════════════════════════════╝"
`;

exports[`createDeferred > strict > resolve reject calls throw 1`] = `
"
╔══════════════════════════════════════════════════╗ RUNNER
d.reject(new Error("test"));
d.promise.catch(() => {
});
d.resolve(1);
╠══════════════════════════════════════════════════╣ OUTPUTS THROW ERROR
Error: Deferred promise already settled
╚══════════════════════════════════════════════════╝"
`;

exports[`createDeferred > strict > resolve resolve calls throw 1`] = `
"
╔══════════════════════════════════════════════════╗ RUNNER
d.resolve(1);
d.resolve(1);
╠══════════════════════════════════════════════════╣ OUTPUTS THROW ERROR
Error: Deferred promise already settled
╚══════════════════════════════════════════════════╝"
`;
